[FSDP][optim_state_dict] Returns the initial states of the empty parameters for KeyedOptimizer/NamedOptimizer

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94130
* #94129

KeyedOptimizer and NamedOptimizer expect the states exist in the state_dict when `load_state_dict` is called even if the corresponding parameters are empty (size == 0). This PR adds the support to make KeyedOptimizer work with `use_orig_params=True`.

Differential Revision: [D43019458](https://our.internmc.facebook.com/intern/diff/D43019458/)

fegin committed:
[FSDP][optim_state_dict] Return the initial states of the empty parameters for KeyedOptimizer/NamedOptimizer

KeyedOptimizer and NamedOptimizer expect the states exist in the state_dict when `load_state_dict` is called even if the corresponding parameters are empty (size == 0). This PR adds the support to make KeyedOptimizer work with `use_orig_params=True`.

Differential Revision: [D43019458](https://our.internmc.facebook.com/intern/diff/D43019458/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Returns the initial states of the empty parameters for KeyedOptimizer/NamedOptimizer"


KeyedOptimizer and NamedOptimizer expect the states exist in the state_dict when `load_state_dict` is called even if the corresponding parameters are empty (size == 0). This PR adds the support to make KeyedOptimizer work with `use_orig_params=True`.

Differential Revision: [D43019458](https://our.internmc.facebook.com/intern/diff/D43019458/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -193,7 +193,9 @@ def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
             # When the conditional training is performed, not all parameters are updated in the optim.
             if param_key not in state.keys():
-                continue
+                raise ValueError(
```
nit: update the comment to indicate that the initial states should be returned?

rohan-varma reviewed (approved):
LGTM overall, but wonder if we can enhance the documentation or otherwise clarify some things here. 

Overall, we have 4 codepaths that are supported: use_orig_params True/False x named_optim True False. 

Could we documented what happens when users use FSDP optim state checkpointing APIs under these configurations? i.e. what are the contracts for non-FSDP managed parameters when do we return no state vs initial / dummy state, do we ever expect eager initialization of optims and what is the different behavior there, etc?

fegin commented:
> LGTM overall, but wonder if we can enhance the documentation or otherwise clarify some things here.
> 
> Overall, we have 4 codepaths that are supported: use_orig_params True/False x named_optim True False.
> 
> Could we documented what happens when users use FSDP optim state checkpointing APIs under these configurations? i.e. what are the contracts for non-FSDP managed parameters when do we return no state vs initial / dummy state, do we ever expect eager initialization of optims and what is the different behavior there, etc?

Yup, sure, I'll enhance the docstring here. I also plan to write down the journey of the integration with TorchRec which can give us the idea what's going of each hack/fix/improvement.

fegin committed:
Update on "[FSDP][optim_state_dict] Returns the initial states of the empty parameters for KeyedOptimizer/NamedOptimizer"


KeyedOptimizer and NamedOptimizer expect the states exist in the state_dict when `load_state_dict` is called even if the corresponding parameters are empty (size == 0). This PR adds the support to make KeyedOptimizer work with `use_orig_params=True`.

Differential Revision: [D43019458](https://our.internmc.facebook.com/intern/diff/D43019458/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fduwjj commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -191,9 +191,14 @@ def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
+            # the assumption of NamedOptimizer.
             if param_key not in state.keys():
-                continue
+                raise ValueError(
```
What about partial training case which enabled in https://github.com/pytorch/pytorch/pull/91344?

fegin committed:
Update on "[FSDP][optim_state_dict] Returns the initial states of the empty parameters for KeyedOptimizer/NamedOptimizer"


KeyedOptimizer and NamedOptimizer expect the states exist in the state_dict when `load_state_dict` is called even if the corresponding parameters are empty (size == 0). This PR adds the support to make KeyedOptimizer work with `use_orig_params=True`.

Differential Revision: [D43019458](https://our.internmc.facebook.com/intern/diff/D43019458/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
